### PR TITLE
0094 cross domain messaging - Supersede Indy HIPE 0022

### PR DESCRIPTION
**Pull request by swcurran**
Updates the HIPE contents to be (more) aligned with the current thinking. In particular, moves from the notion of a fixed structure of agents (domain endpoint and exactly one routing agent/mediator) for an inbound request to using the endpoint DID and routingKeys array in the diddoc. Fixed links in the HIPE that had never been added (my bad).

Ran the check links (and fixed two problems - yay!) and generated index python scripts.

### CODE REVIEW

**Comment by swcurran** [comment]
Added indy-hipe PR to annotate the HIPE content with a superseded message - https://github.com/hyperledger/indy-hipe/pull/137

File: concepts/0094-cross-domain-messaging/README.md [hunk 1]
```diff
@@ -0,0 +1,302 @@
...
+
+The goal of this RFC is to define the rules that domains MUST follow to enable the delivery of Agent messages from a Sending Agent to a Receiver Agent in a secure and privacy-preserving manner.
+
+## Motivation
+
+The purpose of this RFC and its related RFCs is to define a layered messaging protocol such that we can ignore the delivery of messages as we discuss the much richer Agent Messaging types and interactions. That is, we can assume that there is no need to include in an Agent message anything about how to route the message to the Receiver - it just magically happens. Alice (via her App Agent) sends a message to Bob, and (because of implementations based on this series of RFCs) we can ignore how the actual message got to Bob's App Agent.
```
Comment (llorllale) [inline concepts/0094-cross-domain-messaging/README.md]: @swcurran nitpick: I think "we can ignore the **transport** of messages" was a more accurate term, unless I'm missing some context?

File: concepts/0094-cross-domain-messaging/README.md [hunk 2]
```diff
@@ -0,0 +1,302 @@
...
+#### DIDDoc
+
+The term "DIDDoc" is used in this RFC as it is defined in the [DID Specification](https://w3c-ccg.github.io/did-spec/):
+
+- a collection of public keys and endpoints,
+- controlled by an identity,
+- associated with a DID, and
+- used for a relationship.
+
+A DID can be resolved to get its corresponding DIDDoc by any Agent that needs access to the DIDDoc. This is true whether talking about a DID on a Public Ledger, or a pairwise DID persisted only to the parties of the relationship. In the case of pairwise DIDs, it's the (implementation specific) domain's responsibility to ensure such resolution is available to all Agents requiring it.
```
Comment (llorllale) [inline concepts/0094-cross-domain-messaging/README.md]: @swcurran aren't we specifically referring to peer DIDs here? Why not link to that spec?

File: concepts/0094-cross-domain-messaging/README.md [hunk 3]
```diff
@@ -0,0 +1,302 @@
...
+
+#### Messages are Private
+
+Agent Messages sent from a Sender to a Receiver SHOULD be private. That is, the Sender SHOULD encrypt the message with a public key for the Receiver. Any agent in between the Sender and Receiver will know only to whom the message is intended (by DID and possibly keyname within the DID), not anything about the message.
+
+#### The Sender Knows The Receiver
+
+This RFC assumes that the Sender knows the Receiver's DID and, within the DIDDoc for that DID, the keyname to use for the Receiver's Agent. How the Sender knows the DID and keyname to send the message is not defined within this RFC - that is a higher level concern.
+
+The Receiver's DID MAY be a public or pairwise DID, and MAY be on a Public Ledger or a only shared between the parties of the relationship.
```
Comment (llorllale) [inline concepts/0094-cross-domain-messaging/README.md]: @swcurran "or only shared between the parties of the relationship"

File: concepts/0094-cross-domain-messaging/README.md [hunk 4]
```diff
@@ -0,0 +1,302 @@
...
+
+Most Sender-to-Receiver messages will be sent between parties that have shared pairwise DIDs. When that is true, the Sender will (usually) AuthCrypt the message. If that is not the case, or for some other reason the Sender does not want to AuthCrypt the message, AnonCrypt will be used. In either case, the Indy-SDK `pack()` function handles the encryption.
+
+If there are mediators specified in the DID service endpoint for the Receiver agent, the Sender must wrap the message for the Receiver in a 'Forward' message for each mediator. It is assumed that the Receiver can determine the from `did` based on the `to` DID using their pairwise relationship.
+
+```json
+{
+  "@type" : "did:sov:BzCbsNYhMrjHiqZDTUASHg;spec/routing/1.0/forward",
+  "to"   : "verkey of last mediator",
```
Comment (llorllale) [inline concepts/0094-cross-domain-messaging/README.md]: @swcurran according to the text preceding this example, the `to` field should have the did fragment of the Receiver's key

File: concepts/0094-cross-domain-messaging/README.md [hunk 5]
```diff
@@ -0,0 +1,302 @@
...
+The second form is used when the precise key (and hence, the Agent controlling that key) is used to encrypt the Agent Message placed in the `msg` field.
+
+The `msg` field calls the Indy-SDK `pack()` function to encrypt the Agent Message to be forwarded. The Sender calls the `pack()` with the suitable arguments to AnonCrypt or AuthCrypt the message. The `pack()` and `unpack()` functions are described in more detail in the [Encryption Envelope](../../features/0019-encryption-envelope/README.md) RFC.
+
+## Reference
+[reference]: #reference
+
+See the other RFCs referenced in this document:
+
+),  and . Those RFCs should be considered together in understanding  messaging.
```
Comment (llorllale) [inline concepts/0094-cross-domain-messaging/README.md]: @swcurran typo? Maybe the line is unnecessary?

File: concepts/0094-cross-domain-messaging/README.md (hunk 1 above)
Comment (swcurran) [reply concepts/0094-cross-domain-messaging/README.md]: I changed that right at the end from what was in the previous.  I was trying to get away from the notion that it was involved in the transport layer protocols - e.g. http, etc. That's a totally separate issue from this one.  As is stated elsewhere, DIDcomm is agnostic to transport and should work using a variety of protocols, including smoke signals, if carefully generated and equally carefully recorded and converted to bytes.

File: concepts/0094-cross-domain-messaging/README.md (hunk 2 above)
Comment (swcurran) [reply concepts/0094-cross-domain-messaging/README.md]: Yes - the link was later in the doc. I've moved up the link to the peer:did method to the first reference and made sure that each reference to pairwise DID also references the peer:did method.

File: concepts/0094-cross-domain-messaging/README.md (hunk 3 above)
Comment (swcurran) [reply concepts/0094-cross-domain-messaging/README.md]: Doh - thanks! Fixed.

File: concepts/0094-cross-domain-messaging/README.md (hunk 4 above)
Comment (swcurran) [reply concepts/0094-cross-domain-messaging/README.md]: Thanks - reverted. I confused myself on that.

File: concepts/0094-cross-domain-messaging/README.md (hunk 5 above)
Comment (swcurran) [reply concepts/0094-cross-domain-messaging/README.md]: Yes - typo.  Removed.

**Comment by swcurran** [comment]
Updates made per @llorllale comments.

**Pull request merged**